### PR TITLE
[Messenger] Allow to customize message type resolving

### PR DESCRIPTION
**Pull request by l-vo**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

For now messages types (classes names used by the handlers) are resolved in the following ways:
- Class name is included in the body with PHP native serialization
- Class name is in a `type` header with Symfony serializer.

A problem arises when the message comes from another application (potentially not a PHP/Symfony application):
- PHP native serialization can only be used with PHP apps
- Using Symfony serializer, the app that sends the message must use a header name imposed by the consumer and must know the consumer internal structure (the target class name). Unless I miss something, it seems not ideal.

This PR suggests to override this behavior introducing type resolvers.
- Implementing `TypeResolverInterface` allows to resolve a class depending on many headers of the message (that are not necessarily a classname, so the sender doesn't have to know which application will consume the message)
- To go further, `DecodedAwareTypeResolverInterface` allows to resolve a class even without headers, based on the message decoded body.

## Exemple with non-class header value

Given a message with a header `type` (any other header name can be used) with value `mytype` is received:
```yaml
# config/packages/messenger.yaml
framework:
    #...
    messenger:
        serializer:
            default_serializer: messenger.transport.symfony_serializer
            symfony_serializer:
                format: json
                context:
                    type_resolver: 'App\Serializer\TypeResolver'
```
```php
namespace App\Serializer;

use Symfony\Component\Messenger\Transport\Serialization\TypeResolver\TypeResolverInterface;

class TypeResolver implements TypeResolverInterface
{
    //... 
   
    public function resolve(array $encodedEnvelope): string
    {
        $type = $encodedEnvelope['headers']['type'] ?? '';
        $deserializationType = /* Some stuff to transform 'mytype' to a class name to use in deserialization process (or throw a MessageDecodingFailedException) */;

        return $deserializationType;
    }    
}
```

## Exemple without headers usage

```json
{
  "user_id": 4711,
  "newsletter_id": "new-updates",
  "status": "subscribed"
}
```
Given the message body above is received:
```yaml
# config/packages/messenger.yaml
framework:
    #...
    messenger:
        serializer:
            default_serializer: messenger.transport.symfony_serializer
            symfony_serializer:
                format: json
                context:
                    type_resolver: 'App\Serializer\TypeResolver'
```
```php
namespace App\Serializer;

use  App\Message\NewsletterMessage;
use Symfony\Component\Messenger\Transport\Serialization\TypeResolver\DecodedAwareTypeResolverInterface;

class TypeResolver implements DecodedAwareTypeResolverInterface
{
    //... 
   
    public function resolve(array $encodedEnvelope, array $body): string
    {
        if (isset($body['newsletter_id']) {
            return NewsletterMessage::class;
        }
         
        // Some stuff to process other cases or throw a MessageDecodingFailedException
        // Headers from $encodedEnvelope can also be used like in the first example
    }    
}
```

### CODE REVIEW

**Comment by Nyholm** [comment]
Thank you for this PR. 

I also have this scenario when php serializer is not enough for me. I took a similar approach to you and implemented a new serializer that took "non-class name" metadata from my messages. 

I like the idea of having a "out of the box" solution to this scenario. But Im not sure we should modify the existing serializer but rather create a new one. 

I've created a package for this: https://github.com/Happyr/message-serializer

**Comment by l-vo** [comment]
@Nyholm Thank you for you feedback 🙂

What have you in mind about creating a new serializer please ? If new serializer there is, it should rely on the Symfony serializer anyway; so it would be very similar to the current `Symfony\Component\Messenger\Transport\Serialization\Serializer`; wouldn't it ?

**Comment by Nyholm** [comment]
Have a look at the package I linked. It uses similar principles as you are doing and it solves the exact same problem.

**Comment by l-vo** [comment]
@Nyholm Yes I saw the package you linked, indeed it solves the same problem but not in the same way.
- The bundle chooses to totally let to the application the implementation of normalization and denormalization. 
- My approach is a bit different, first, I don't care about serialization (and it's probably ok that a bundle has a wider action range than the core code). Secondly IMHO it would be a shame to not use the serializer component, it provides all that we need, the only thing that is missing here is to known the target class for denormalization.

Of course it's simply my opinion, if people think it's better to use a *transformer/hydrator* approach, I have no problem with changing my implementation 🙂

@sroze have you any point of view about this please ?

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 1]
```diff
@@ -62,12 +69,12 @@ public static function create(): self
      */
     public function decode(array $encodedEnvelope): Envelope
     {
-        if (empty($encodedEnvelope['body']) || empty($encodedEnvelope['headers'])) {
-            throw new MessageDecodingFailedException('Encoded envelope should have at least a "body" and some "headers".');
-        }
-
-        if (empty($encodedEnvelope['headers']['type'])) {
-            throw new MessageDecodingFailedException('Encoded envelope does not have a "type" header.');
+        $headersRequired = !isset($this->context[self::TYPE_RESOLVER]) || $this->context[self::TYPE_RESOLVER] instanceof HeaderTypeResolver;
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:72]: This looks weird to me - we're hardcoding a special case for `HeaderTypeResolver`. Is this the exception needed anymore? Or could we fail *later* with some message about needing body/headers or a type resolver?

File: src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/HeaderTypeResolver.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * @author Laurent VOULLEMIER <laurent.voullemier@gmail.com>
+ */
+final class HeaderTypeResolver implements TypeResolverInterface
+{
+    /**
+     * @var string
+     */
+    private $field;
+
+    public function __construct(string $field = 'type')
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/HeaderTypeResolver.php:28]: Maybe `$headerName` ?

File: src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/HeaderTypeResolver.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Resolve denormalization class based on a message header value.
+ *
+ * @author Laurent VOULLEMIER <laurent.voullemier@gmail.com>
+ */
+final class HeaderTypeResolver implements TypeResolverInterface
+{
+    /**
+     * @var string
+     */
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/HeaderTypeResolver.php:25]: phpdoc is not needed

File: src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/TypeResolverInterface.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+namespace Symfony\Component\Messenger\Transport\Serialization\TypeResolver;
+
+/**
+ * Resolve denormalization class based on message headers.
+ *
+ * @author Laurent VOULLEMIER <laurent.voullemier@gmail.com>
+ */
+interface TypeResolverInterface
+{
+    public function resolve(array $encodedEnvelope): string;
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/TypeResolverInterface.php:21]: The method should have phpdoc that describes what it should return I think

**Review by weaverryan — COMMENTED** [review]
I like this. It replaces the "hardcoded" logic of looking on the `type` header for the type with an extensible system that defaults to this behavior.

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php (hunk 1 above)
Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: It was to stay close to the original behavior. But a change in the exception message can probably not be considered as a BC break :)

**Comment by l-vo** [comment]
@weaverryan Thank you for your review. All threads are resolved.

**Comment by fabpot** [comment]
@l-vo Can you have a look at the failing tests?

**Comment by l-vo** [comment]
@fabpot Tests fixed :)

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 5]
```diff
@@ -79,14 +82,38 @@ public function decode(array $encodedEnvelope): Envelope
...
+        $resolver = $context[self::TYPE_RESOLVER] ?? new HeaderTypeResolver();
+
+        if ($resolver instanceof TypeResolverInterface) {
+            $type = $resolver->resolve($encodedEnvelope);
+
+            return $this->serializer->deserialize($encodedEnvelope['body'], $type, $this->format, $context);
+        }
+
+        if ($resolver instanceof DecodedAwareTypeResolverInterface) {
+            if (!$this->serializer instanceof DecoderInterface || !$this->serializer instanceof DenormalizerInterface) {
```
Comment (Nyholm) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:104]: This is a BC break. 

If I currently pass a `Symfony\Component\Serializer\SerializerInterface` as first argument to the `Symfony\Component\Messenger\Transport\Serialization\Serializer`, then it will stop working if we merge this PR.

**Review by Nyholm — CHANGES_REQUESTED** [review]
The PR contains a BC break so it is a big 👎  at this state. 

I like that you are trying to make it easier to solve the problem of deserialising messages from a different application. Could you update the PR description with some code that shows how this feature could be used? Say that the following json message comes in form the queue, what config and classes am I required to have?

```json
{
  "user_id": 4711,
  "newsletter_id": "new-updates",
  "status": "subscribed"
}
```

I am also not sure if we should handle the problem of "receive messages from other app" without discussing the very related problem of "send messages to other app".

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php (hunk 5 above)
Comment (Nyholm) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:104]: Hm.. I might be wrong. Since the default resolver is a `TypeResolverInterface` this code will never be reached.

File: src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/TypeResolverInterface.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Serialization\TypeResolver;
+
+/**
+ * Resolve denormalization class based on message headers.
```
[the remark this answers is not in the document]
Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/TypeResolverInterface.php:15]: Yes the body could be used BUT when using this interface the body is encoded. So `DecodedAwareTypeResolverInterface` is more suitable to deal with the body.

File: src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/HeaderTypeResolver.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    public function __construct(string $headerName = 'type')
+    {
+        $this->headerName = $headerName;
+    }
+
+    public function resolve(array $encodedEnvelope): string
+    {
+        $headers = $encodedEnvelope['headers'] ?? [];
+        if (!isset($headers[$this->headerName]) || !$headers[$this->headerName]) {
```
[the remark this answers is not in the document]
Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/TypeResolver/HeaderTypeResolver.php:33]: Yes it is, but I seem to remember that Fabien told me in a review (I can't find this review) to not use `empty` in core.

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 8]
```diff
@@ -79,14 +82,38 @@ public function decode(array $encodedEnvelope): Envelope
...
+    {
+        $resolver = $context[self::TYPE_RESOLVER] ?? new HeaderTypeResolver();
+
+        if ($resolver instanceof TypeResolverInterface) {
+            $type = $resolver->resolve($encodedEnvelope);
+
+            return $this->serializer->deserialize($encodedEnvelope['body'], $type, $this->format, $context);
+        }
+
+        if ($resolver instanceof DecodedAwareTypeResolverInterface) {
```
[the remark this answers is not in the document]
Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:103]: Maybe I misunderstand what you would like, but I think it's not possible since `TypeResolverInterface` and `DecodedAwareTypeResolverInterface` have no common ancestor.

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php (hunk 5 above)
Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:104]: Yes indeed, this occurs only with `DecodedAwareTypeResolverInterface` that is not the default resolver.

**Comment by l-vo** [comment]
status: needs review

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (Nyholm) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:103]: Rewrite this line as `if (!$resolver instanceof DecodedAwareTypeResolverInterface) {` and change the body of the if-statement

[the remark this answers is not in the document]
Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:103]: Ah ok :) Updated

**Comment by l-vo** [comment]
@Nyholm 

> I am also not sure if we should handle the problem of "receive messages from other app" without discussing the very related problem of "send messages to other app".

I confess I didn't think about the problem to send message to an outside app. But (just a quick idea) I guess it would be possible via a kind of `ExtraHeadersStamp` wouldn't it ? AFAIK the only problem to send to another app (non-sf) is only to be able to add specifics headers ? Or is there any other stuff that I miss ?

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 9]
```diff
@@ -79,14 +82,38 @@ public function decode(array $encodedEnvelope): Envelope
...
+        $resolver = $context[self::TYPE_RESOLVER] ?? new HeaderTypeResolver();
+
+        if ($resolver instanceof TypeResolverInterface) {
+            $type = $resolver->resolve($encodedEnvelope);
+
+            return $this->serializer->deserialize($encodedEnvelope['body'], $type, $this->format, $context);
+        }
+
+        if (!$resolver instanceof DecodedAwareTypeResolverInterface) {
+            throw new MessageDecodingFailedException(sprintf(sprintf('"%s" must be an instance of "%s" or "%s", "%s" given.', self::TYPE_RESOLVER, TypeResolverInterface::class, DecodedAwareTypeResolverInterface::class, \is_object($resolver) ? \get_class($resolver) : \gettype($resolver))));
```
Comment (Nyholm) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:104]: It looks like you have two `sprintf` nested here. That should be fixed.

Comment (l-vo) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: Indeed :-s. Thank you 🙂

**Comment by Nyholm** [comment]
Thank you @l-vo. I think the implementation looks better now. 

I still have some doubts. Here is what's on my mind. 

### Context

It is currently allowed for a message to have a `SerializerStamp` with some context. That context is used to select a `TypeResolver`. Since this PR is about making sure other applications shouldn't be aware of our (the consuming application) internal classes, I dont think it is a good idea to allow that. 

I am saying that maybe the `context` isn't the place so specify a type resolver. 

### The alternative approach

Thank you for updating the PR description. It makes it easier to understand. 

Currently, the alternative approach for this PR is to do something like this: 

```yaml
# config/packages/messenger.yaml
framework:
    #...
    messenger:
        serializer:
            default_serializer: 'App\Serializer\MyMessengerSerializer'

services:
    App\Serializer\MyMessengerSerializer: 
        arguments: ['@Symfony\Component\Messenger\Transport\Serialization\Serializer']
```

```php
namespace App\Serializer;

use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;

class MyMessengerSerializer implements SerializerInterface
{
    private $inner;

    public function __construct(SerializerInterface $inner) {
        $this->inner = $inner;
    }

    public function encode(Envelope $envelope): array
    {
        return $this->inner->encode($envelope);
    }

    public function decode(array $encodedEnvelope): Envelope
    {
         $externalSystemMessage = $encodedEnvelope['body'];
         // Do whatever business logic you want here. 
         
         if (/* is not from external system */ ) {
             return $this->inner->decode($encodedEnvelope);
         }
         $message = new MyMessage(json_decode($externalSystemMessage));

         return new Envelope($message);
    }
}
```

I dont see this PR as a massive win. I do think it is a step towards "making things a bit easier". However, I am not sure this is the correct way to make things a bit easier. 

One could imagine a PR that introduces a third serializer to the messenger component. It can take a fallback serializer and a TypeResolver as arguments. That way we avoid the `context` issue and it will give you all the flexibility in the world. 

One could also imagine a [SerializerRouter](https://github.com/Happyr/message-serializer/blob/0.5.0/src/SerializerRouter.php) that choose the correct serializer depending on the format. 


------------

The bottom line is: 
- It is easy enough to add a custom serializer. 
- I do believe we can improve when dealing with external incoming and outgoing messages.
- I dont think patching the existing Serializer like this is the one solution we want to move forward with. 

Third party bundles and libraries are easy to jack into the serialising process with just a line of configuration. 

```diff
# config/packages/messenger.yaml

 framework:
     messenger:
         transports:            
             to_external:
               dsn: '%env(MESSENGER_EXTERNAL)%'
+              serializer: 'Happyr\MessageSerializer\Serializer'
```

I do see value in this implementation. But I dont think we should add it to the core. It will do well as a separate package. 

I am still 👎

**Comment by fabpot** [comment]
Closing per @Nyholm detailed comment. Thanks for proposing.
